File: solutions/206-reverse-linked-list/solution.py
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def reverse_list(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head is None or head.next is None:
            return head
        prev = None
        current = head
        while current is not None:
            nxt = current.next
            current.next = prev
            prev = current
            current = nxt
        return prev


if __name__ == "__main__":
    import sys
    from pathlib import Path
    sys.path.insert(0, str(Path(__file__).resolve().parents[1] / "_helpers"))
    from testutil import build_list, list_to_slice
    sol = Solution()
    head = build_list([1, 2, 3, 4, 5])
    got = list_to_slice(sol.reverse_list(head))
    status = "PASS" if got == [5, 4, 3, 2, 1] else "FAIL"
    print(f"{status} | reversed={got}")

File: solutions/206-reverse-linked-list/solution.go
package main

import "fmt"

type ListNode struct {
	Val  int
	Next *ListNode
}

func sliceToList(nums []int) *ListNode {
	dummy := &ListNode{}
	cur := dummy
	for _, v := range nums {
		cur.Next = &ListNode{Val: v}
		cur = cur.Next
	}
	return dummy.Next
}

func listToSlice(head *ListNode) []int {
	out := []int{}
	for head != nil {
		out = append(out, head.Val)
		head = head.Next
	}
	return out
}

func reverseList(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}
	var prev *ListNode
	current := head
	for current != nil {
		next := current.Next
		current.Next = prev
		prev = current
		current = next
	}
	return prev
}

func main() {
	head := sliceToList([]int{1, 2, 3, 4, 5})
	got := listToSlice(reverseList(head))
	want := []int{5, 4, 3, 2, 1}
	status := "PASS"
	for i := range want {
		if got[i] != want[i] {
			status = "FAIL"
		}
	}
	fmt.Printf("%s | reversed=%v\n", status, got)
}
